panther_hardware_interfaces/src/panther_system.cpp

```diff
@@ -0,0 +1,519 @@
+    }
+
+    if (joint.state_interfaces[1].name != hardware_interface::HW_IF_VELOCITY) {
+      throw std::runtime_error(
+        "Joint "s + joint.name + " have "s + joint.state_interfaces[1].name +
+        " as second state interface. "s + hardware_interface::HW_IF_VELOCITY + " expected."s);
+    }
+
+    if (joint.state_interfaces[2].name != hardware_interface::HW_IF_EFFORT) {
+      throw std::runtime_error(
+        "Joint "s + joint.name + " have "s + joint.state_interfaces[2].name +
+        " as third state interface. "s + hardware_interface::HW_IF_EFFORT + " expected."s);
+    }
+  }
+}
+
+void PantherSystem::ReadDrivetrainSettings()
+{
+  drivetrain_settings_.motor_torque_constant =
+    std::stof(info_.hardware_parameters["motor_torque_constant"]);
+  drivetrain_settings_.gear_ratio = std::stof(info_.hardware_parameters["gear_ratio"]);
+  drivetrain_settings_.gearbox_efficiency =
+    std::stof(info_.hardware_parameters["gearbox_efficiency"]);
+  drivetrain_settings_.encoder_resolution =
+    std::stof(info_.hardware_parameters["encoder_resolution"]);
+  drivetrain_settings_.max_rpm_motor_speed =
+    std::stof(info_.hardware_parameters["max_rpm_motor_speed"]);
+}
+
+void PantherSystem::ReadCanOpenSettings()
+{
+  canopen_settings_.can_interface_name = info_.hardware_parameters["can_interface_name"];
+  canopen_settings_.master_can_id = std::stoi(info_.hardware_parameters["master_can_id"]);
+  canopen_settings_.front_driver_can_id =
+    std::stoi(info_.hardware_parameters["front_driver_can_id"]);
+  canopen_settings_.rear_driver_can_id = std::stoi(info_.hardware_parameters["rear_driver_can_id"]);
+  canopen_settings_.pdo_feedback_timeout =
+    std::chrono::milliseconds(std::stoi(info_.hardware_parameters["pdo_feedback_timeout"]));
+  canopen_settings_.sdo_operation_timeout =
+    std::chrono::milliseconds(std::stoi(info_.hardware_parameters["sdo_operation_timeout"]));
+}
+
+void PantherSystem::ReadInitializationActivationAttempts()
+{
+  max_roboteq_initialization_attempts_ =
+    std::stoi(info_.hardware_parameters["max_roboteq_initialization_attempts"]);
+  max_roboteq_activation_attempts_ =
+    std::stoi(info_.hardware_parameters["max_roboteq_activation_attempts"]);
+  max_safety_stop_attempts_ = std::stoi(info_.hardware_parameters["max_safety_stop_attempts"]);
+}
+
+void PantherSystem::ReadParametersAndCreateRoboteqErrorFilter()
+{
+  const unsigned max_write_sdo_errors_count =
+    std::stoi(info_.hardware_parameters["max_write_sdo_errors_count"]);
+  const unsigned max_read_sdo_errors_count =
+    std::stoi(info_.hardware_parameters["max_read_sdo_errors_count"]);
+  const unsigned max_read_pdo_errors_count =
+    std::stoi(info_.hardware_parameters["max_read_pdo_errors_count"]);
+
+  roboteq_error_filter_ = std::make_shared<RoboteqErrorFilter>(std::vector<ErrorFilter>{
+    ErrorFilter(max_read_sdo_errors_count), ErrorFilter(max_write_sdo_errors_count),
+    ErrorFilter(max_read_pdo_errors_count), ErrorFilter(1)});
+}
```

Comment:
```suggestion
  roboteq_error_filter_ = std::make_shared<RoboteqErrorFilter>(std::vector<ErrorFilter>{
    ErrorFilter(max_read_sdo_errors_count, READ_SDO), ErrorFilter(max_write_sdo_errors_count, WRITE_SDO),
    ErrorFilter(max_read_pdo_errors_count, READ_PDO), ErrorFilter(1, ROBOTEQ_DRIVER)});
}
```
Don't like the way this is initialized and we hope it is in right order with enum definition. It will be better to pass some unique ID when definning ErrorFilter, this way you don't have to worry how you pass filters.
Or you could define error filters ealier and access them with `GetID` method

Reply:
yeah, that wasn't the best approach, I modified it, please check out this commit: https://github.com/husarion/panther_ros/pull/208/commits/8afec1e9a847fb33311bb8ecd2b1cae9e8d2ceff